torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -3085,8 +3078,39 @@ def _wait_for_post_backward(self) -> None:
         # A backward pass is done, clean up below.
         self._exec_order_data.reset()
 
+        def _catch_all_reshard(fsdp_module: FullyShardedDataParallel) -> None:
+            """
+            Reshards full parameters that may have not been resharded in
+            post_backward_hook. This can happen when an FSDP module's output
+            is used in forward so its pre-backward fires unsharding the param,
+            but post-backward does not fire since the output was not ultimately
+            used in loss computation so FSDP parameter did not get a gradient.
+            """
+            try:
+                for p in fsdp_module.params:
+                    # Parameter is already resharded if the param in use points
+                    # to the local shard.
+                    if p.data.data_ptr() == p._local_shard.data_ptr():
```

Comment:
nit: Like the previous comment, I think `p.data.data_ptr()` may possibly be simplified to `p.data_ptr()`.

Mostly as a note to myself:
This check is a bit dangerous. This carries the assumption that if we switched to using the local shard, then we must have correctly freed the unsharded parameter if needed. Given our implementation, this should be correct, but I am concerned about this from a design perspective. If something changes outside of this catch-all reshard to break the assumption, then this will error silently. I will think about this more.

Reply:
filed an issue for this https://github.com/pytorch/pytorch/issues/83956